pkg/controller/util/endpoint/controller_utils_test.go

```diff
@@ -224,3 +231,171 @@ func TestShouldPodBeInEndpoints(t *testing.T) {
+			if err != nil {
+				t.Errorf("Error from cache.GetPodServiceMemberships: %v", err)
+			} else if !services.Equal(test.expect) {
+				t.Errorf("Expect service %v, but got %v", test.expect, services)
+			}
+		})
+	}
+}
+
+func TestServiceSelectorCache_Update(t *testing.T) {
+	var selectors []labels.Selector
+	for i := 0; i < 5; i++ {
+		selector := labels.Set(map[string]string{"app": fmt.Sprintf("test-%d", i)}).AsSelectorPreValidated()
+		selectors = append(selectors, selector)
+	}
+	tests := []struct {
+		name   string
+		key    string
+		cache  *ServiceSelectorCache
+		update map[string]string
+		expect labels.Selector
+	}{
+		{
+			name:   "add test/service-0",
+			key:    "test/service-0",
+			cache:  generateServiceSelectorCache(map[string]labels.Selector{}),
+			update: map[string]string{"app": "test-0"},
+			expect: selectors[0],
+		},
+		{
+			name:   "add test/service-1",
+			key:    "test/service-1",
+			cache:  generateServiceSelectorCache(map[string]labels.Selector{"test/service-0": selectors[0]}),
+			update: map[string]string{"app": "test-1"},
+			expect: selectors[1],
+		},
+		{
+			name:   "update test/service-2",
+			key:    "test/service-2",
+			cache:  generateServiceSelectorCache(map[string]labels.Selector{"test/service-2": selectors[2]}),
+			update: map[string]string{"app": "test-0"},
+			expect: selectors[0],
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			selector := test.cache.Update(test.key, test.update)
+			if !reflect.DeepEqual(selector, test.expect) {
+				t.Errorf("Expect selector %v , but got %v", test.expect, selector)
+			}
+		})
+	}
+}
+
+func generateServiceSelectorCache(cache map[string]labels.Selector) *ServiceSelectorCache {
+	return &ServiceSelectorCache{
+		cache: cache,
+	}
+}
+
+func BenchmarkGetPodServiceMemberships(b *testing.B) {
+	// init fake service informer.
+	fakeInformerFactory := informers.NewSharedInformerFactory(&fake.Clientset{}, 0*time.Second)
+	for i := 0; i < 1000; i++ {
+		service := &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("service-%d", i),
+				Namespace: "test",
+			},
+			Spec: v1.ServiceSpec{
+				Selector: map[string]string{
+					"app": fmt.Sprintf("test-%d", i),
+				},
+			},
+		}
+		fakeInformerFactory.Core().V1().Services().Informer().GetStore().Add(service)
+	}
+
+	pod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-pod-0",
+			Labels: map[string]string{
+				"app": "test-0",
+			},
+		},
+	}
+
+	cache := NewServiceSelectorCache()
+	expect := sets.NewString("test/service-0")
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		services, err := cache.GetPodServiceMemberships(fakeInformerFactory.Core().V1().Services().Lister(), pod)
+		if err != nil {
+			b.Fatalf("Error from GetPodServiceMemberships(): %v", err)
+		}
+		if !reflect.DeepEqual(services, expect) {
```

Comment:
The benchmark isn't doing exactly what you want - DeepEqual is relatively expensive and may potentially influence the results. I would recommend that checking the size or sth like that.

Also in the PR desc, please paste the benchmark results (simply run sth like:
`go test k8s.io/kubernetes/pkg/controller/utils/endpoint -test.run=XXXXX -bench=Membership -benchmem`

Reply:
addressed, quite a difference after that